Apply zero traces to the SMT

Stefan-Ethernal opened the pull request:
Closes #1369

Stefan-Ethernal committed:
feat: implement ApplyTraces for SMT (initial approach)

Stefan-Ethernal committed:
Handle self destructed in apply traces

Stefan-Ethernal committed:
Fix ReadAccountStorage in the SMT

Stefan-Ethernal committed:
SMT implements StateWriter and left some TODO (open questions)

Stefan-Ethernal committed:
Copy SMT

cffls commented on smt/pkg/smt/smt_state_writer.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	// TODO: @Stefan-Ethernal How to update:
+	// 1. the account bytecode (we have CodeHash, but not the Code itself, so how to retrieve the actual Code) and
+	// 2. the storage slots (we only have the Root, but not the storage slots themselves)
```
We don't need to update code or storage slot, which will be updated in other functions below.

cffls commented on smt/pkg/smt/smt_state_writer.go:
```diff
@@ -0,0 +1,140 @@
+	keyNonce := utils.KeyEthAddrNonce(hexAddr)
+	keyContractCode := utils.KeyContractCode(hexAddr)
+
+	// TODO: @Stefan-Ethernal Remove contract storage, but how to retrieve the contract storage slots,
```
We don't need to remove the storage slots. I doubt this function will be used because an account can't be deleted after Cancun.

Stefan-Ethernal replied:
Ok, in that case, I have left it unimplemented and just returned an error https://github.com/0xPolygonHermez/cdk-erigon/pull/1398/commits/6abc3e1dd25d203de4fc2237a1a2b2d84f1318d7.

IMO, it is better to have it unimplemented, than half implemented.

cffls commented on smt/pkg/smt/smt_state_writer.go:
```diff
@@ -0,0 +1,140 @@
+		return nil
+	}
+
+	// TODO: @Stefan-Ethernal Check is this correct? Hopefully it would not overwrite the entire (already existing) storage.
```
Seems okay, if the a already exists in SMT, it should be overwritten.

Stefan-Ethernal committed:
Remove TODOs and minor changes

Stefan-Ethernal committed:
Handle OpSMTLeaf case in witness deserialization

Stefan-Ethernal committed:
Change letter case

Stefan-Ethernal committed:
ApplyTraces UT (initial approach WIP)

Stefan-Ethernal committed:
Restructure and add empty block test case

Stefan-Ethernal committed:
Pass the empty block apply traces test

cffls committed:
Update test using a full witness

Stefan-Ethernal committed:
Simplify TestOperatorSMTLeafValue_EncodeDecode

Stefan-Ethernal committed:
Rename test cases

Stefan-Ethernal committed:
Handle errors in the smt.insert

Stefan-Ethernal committed:
Optimize BuildSMTFromWitness (capture result in local vars)

Stefan-Ethernal committed:
Rename BuildSMTFromWitness to match capitalization

Stefan-Ethernal committed:
Rename test cases, introduce isFullWitness flag and pass the empty block test case

Stefan-Ethernal committed:
SC deployment and interaction test vectors (WIP)

Stefan-Ethernal committed:
Some simplifications in insertHashNode function

Stefan-Ethernal committed:
Rebase and fix TestSMTApplyTraces

Stefan-Ethernal committed:
Use SetAccountState for setting balance and nonce when applying zero trace

IvanBelyakoff committed:
Fix 'full-witness-contract-interaction.json' test (#1558)

Set balance and nonce separately, as one might be missing
which leads to invalid state.
Set bytecode only when Write is not empty.
